_posts/2019-11-25-query-pulsar-streams-using-apache-flink.md

```diff
@@ -0,0 +1,220 @@
+// ingest  DataStream with Pulsar consumer
+DataStream<String> words = env.addSource(src);
+```
+
+Pulsar streams can then get connected to the Flink processing logic…
+
+```java
+// perform computation on DataStream (here a simple WordCount)
+DataStream<WordWithCount> wc = words
+  .flatmap((FlatMapFunction<String, WordWithCount>) (word, collector) -> {
+    collector.collect(new WordWithCount(word, 1));
+  })
+ 
+  .returns(WordWithCount.class)
+  .keyBy("word")
+  .timeWindow(Time.seconds(5))
+  .reduce((ReduceFunction<WordWithCount>) (c1, c2) ->
+    new WordWithCount(c1.word, c1.count + c2.count));
+```
+
+...and then get emitted back to Pulsar (used now as a sink), sending one’s computation results downstream, back to a Pulsar topic: 
+
+
+```java
+// emit result via Pulsar producer 
+wc.addSink(new FlinkPulsarProducer<>(
+  serviceUrl,
+  outputTopic,
+  new AuthentificationDisabled(),
+  wordWithCount -> wordWithCount.toString().getBytes(UTF_8),
+  wordWithCount -> wordWithCount.word)
+);
+```
+
+Although this is a great first integration step, the existing design is not leveraging the full power of the framework. Some shortcomings of the integration with Flink 1.6.0 relate to Pulsar neither being utilized as durable storage nor having schema integration with Flink, resulting in manual input when describing an application’s schema registry.
+
+
+# Pulsar’s integration with Flink 1.9: Using Pulsar as a Flink catalog
+
+The latest integration between [Flink 1.9.0](https://flink.apache.org/downloads.html#apache-flink-191) and Pulsar addresses most of the previously mentioned shortcomings. The contribution of Alibaba’s Blink to the Flink repository adds many enhancements and new features to the processing framework that make the integration with Pulsar significantly more powerful and impactful. Flink 1.9.0 brings Pulsar schema integration into the picture, makes the Table API a first-class citizen and provides an exactly-once streaming source and at-least-once streaming sink with Pulsar. Lastly, with schema integration, Pulsar can now be registered as a Flink catalog, making running Flink queries on top of Pulsar streams a matter of a few commands. In the following sections, we will take a closer look at the new integrations and provide examples of how to query Pulsar streams using Flink SQL.
+
+
+# Leveraging the Flink <> Pulsar Schema Integration
+
+Before delving into the integration details and how you can use Pulsar schema with Flink, let us describe how schema in Pulsar works. Schema in Apache Pulsar already co-exists and represents the consensus of data on the broker side of the framework, something that makes schema registry with external systems obsolete. Additionally, the data schema in Pulsar is associated with each topic so both producers and consumers send data with predefined schema information, while the broker performs schema validation, and manages schema multi-versioning and evolution in compatibility checks. 
```

Comment:
What do you mean by "consensus of data"?

Reply:
here it's more the "representation of the data", should I update it?